apps/mymedicare_cb/models.py

```diff
@@ -3,68 +3,89 @@
+        email
     Returns:
         A User
-
     Raises:
         KeyError: If an expected key is missing from user_info.
         KeyError: If response from fhir server is malformed.
         AssertionError: If a user is matched but not all identifiers match.
     """
-    subject = user_info['sub']
-    hicn = user_info['hicn']
-    # Convert SLS's mbi to UPPER case.
-    mbi = user_info['mbi'].upper()
-
-    # If mbi is empty set to None
-    if mbi == "":
-        mbi = None
-
-    # Create hashed values.
-    hicn_hash = hash_hicn(hicn)
-    mbi_hash = hash_mbi(mbi)
-
-    # TODO: Add return for ID type used for FHIR_ID match.
-    # raises exceptions.NotFound:
-    fhir_id, backend_data = match_hicn_hash(hicn_hash)
-
-    # HICN was used to match the user.
-    user_id_type = 'H'
+    # Match a patient identifier via the backend FHIR server
+    fhir_id, hash_lookup_type = match_fhir_id(mbi_hash=mbi_hash, hicn_hash=hicn_hash)
 
     try:
+        # Does an existing user and crosswalk exist for SLS username?
         user = User.objects.get(username=subject)
+
+        # TODO: Replace asserts with exception handling.
+        if user.crosswalk.user_hicn_hash != hicn_hash:
+            mesg = "Found user's hicn did not match"
+            log_get_and_update_user(logger, user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, mesg)
         assert user.crosswalk.user_hicn_hash == hicn_hash, "Found user's hicn did not match"
+
+        if user.crosswalk.fhir_id != fhir_id:
+            mesg = "Found user's fhir_id did not match"
+            log_get_and_update_user(logger, user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, mesg)
         assert user.crosswalk.fhir_id == fhir_id, "Found user's fhir_id did not match"
+
+        if user.crosswalk.user_mbi_hash is not None:
+            if user.crosswalk.user_mbi_hash != mbi_hash:
+                mesg = "Found user's mbi did not match"
+                log_get_and_update_user(logger, user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, mesg)
+            assert user.crosswalk.user_mbi_hash == mbi_hash, "Found user's mbi did not match"
+        else:
+            # Previously stored value was None/Null and mbi_hash != None, update just the mbi hash.
+            if mbi_hash is not None:
+                mesg = "UPDATE mbi_hash since previous value was NULL"
+                log_get_and_update_user(logger, user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, mesg)
+                user.crosswalk.user_mbi_hash = mbi_hash
+                user.crosswalk.save()
+
+        # Update hash type used for lookup, if it has changed from last match.
+        if user.crosswalk.user_id_type != hash_lookup_type:
+            mesg = "UPDATE user_id_type as it has changed from the previous lookup value"
+            log_get_and_update_user(logger, user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, mesg)
+            user.crosswalk.user_id_type = hash_lookup_type
+            user.crosswalk.save()
```

Comment:
Does this run in a DB transaction with the above DB operation, or separately?

If separately and your code unexpectedly went boom between them, would your DB be left in an inconsistent & unrecoverable state for the user?

Reply:
> Does this run in a DB transaction with the above DB operation, or separately?
> 
> If separately and your code unexpectedly went boom between them, would your DB be left in an inconsistent & unrecoverable state for the user?

That could be an issue related to that hash_lookup_type not getting updated.

Further, in the case where `user.crosswalk.user_mbi_hash = None`, it will perform two separate model/db `.save()`  operations.

I'm changing to add a `user.crosswalk.user_id_type = hash_lookup_type` line previous to the 1st `.save()`.  This change will improve it to only perform a single `.save()` operation for that case---instead of two.

Reply:
Is there a way to wrap the whole thing in a single DB transaction? (Sorry, not familiar enough with Django to know, myself.)

Might make it easier to reason about, I think.

Reply:
@karlmdavis Those `user.crosswalk.save()` method calls are for updating an existing Crosswalk model instance in the DB.  If the User/Crosswalk doesn't exist yet, the code in the `create_beneficiary_record()` function further down has the creates for the related User/UserProfile/Crosswalk model instances in a transaction via the `with transaction.atomic()` context.